lib/sage_accounting_api_client.php

```diff
@@ -0,0 +1,181 @@
+  }
+
+  /* POST request to renew the access_token */
+  public function renewAccessToken() {
+    $params = array("client_id" => $this->clientId,
+                    "client_secret" => $this->clientSecret,
+                    "refresh_token" => $this->getRefreshToken(),
+                    "grant_type" => "refresh_token");
+
+    $response = $this->getToken($params);
+
+    return $this->storeAccessToken($response);
+  }
+
+  /* GET request */
+  public function execApiRequest($resource, $httpMethod, $postData = NULL) {
+    $curl = $this->prepareApiRequest($resource, $httpMethod, $postData);
+    $response = $this->sendApiRequest($curl);
+
+    return $this->buildApiResponse($curl, $response);
+  }
+
+  /**
+  * Returns the previously loaded access token
+  */
+  public function getAccessToken() {
+    return $this->getAccessTokenStore()->getAccessToken();
+  }
+
+  /**
+  * Returns the previously loaded UNIX timestamp when the access token expires
+  */
+  public function getExpiresAt() {
+    return $this->getAccessTokenStore()->getExpiresAt();
+  }
+
+  /**
+  * Returns the previously loaded refresh token
+  */
+  public function getRefreshToken() {
+    return $this->getAccessTokenStore()->getRefreshToken();
+  }
+
+  public function getAccessTokenStore() {
+    if ($this->accessTokenStore) return $this->accessTokenStore;
+
+    $this->accessTokenStore = new AccessTokenStore();
+    if (!$this->accessTokenStore->load()) {
+      $this->accessTokenStore = NULL;
+    }
+
+    return $this->accessTokenStore;
+  }
+
+  // Private area
+
+  private function prepareApiRequest($resource, $method, $postData = NULL) {
+    $endpoint = self::BASE_ENDPOINT . $resource;
+    $method = strtoupper($method);
+
+    $curl = curl_init($endpoint);
+    curl_setopt($curl, CURLOPT_CUSTOMREQUEST, $method);
+    curl_setopt($curl, CURLOPT_HEADER, false);
+    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+    curl_setopt($curl, CURLOPT_HTTPHEADER, $this->getRequestHeaders());
+
+    if ($postData && ($method == 'POST' || $method == 'PUT')) {
+      curl_setopt($curl, CURLOPT_POSTFIELDS, $postData);
+    }
+
+    curl_setopt($curl, CURLINFO_HEADER_OUT, TRUE);
+
+    return $curl;
+  }
+
+  private function sendApiRequest($curl) {
+    $response = curl_exec($curl);
+
+    if (!$response) {
+      /* Handle errors: DNS lookup failed, connection timeout, read timeout */
+    }
+
+    return $response;
+  }
+
+  private function buildApiResponse($curl, $response) {
+    return new SageAccountingApiResponse($curl, $response);
+  }
+
+  private function getToken($params) {
+    $options = array('http' => array('header'  => "Content-type: application/x-www-form-urlencoded",
+                                     'method'  => 'POST',
+                                     'content' => http_build_query($params)));
+    $context  = stream_context_create($options);
+    $result = file_get_contents(self::TOKEN_ENDPOINT, false, $context);
```

Comment:
What do you think about using an OAuth2 client library like https://github.com/thephpleague/oauth2-client ?